File: Python projects/Python Fundamentals/Final_exam/01_Programming_Fundamentals_Final_Exam_Retake/02_Ad_Astra.py
import re

text = input()
pattern = r"(#|\|)(?P<name>[a-zA-Z\s]+)\1(?P<day>\d{2})/(?P<month>\d{2})/(?P<year>\d{2})\1(?P<calories>\d+)\1"

matches = re.finditer(pattern, text)

# print([match.groupdict() for match in matches])
foods= [match.groupdict()for match in matches]
matches = re.finditer(pattern, text)
sum_calories = 0
for match in matches:
    sum_calories += int(match.group(6))
print(f"You have food to last you for: {sum_calories// 2000} days!")

for food in foods:
    print(f"Item: {food['name']}, Best before: {food['day']}/{food['month']}/{food['year']}, Nutrition: {food['calories']}")


'''
TASK:
You are an astronaut who just embarked on a mission across the solar system. Since you will be in space for a long time, 
you have packed a lot of food with you. Create a program, which helps you identify how much food you have left and gives 
you information about its expiration date.
On the first line of the input, you will be given a text string. You must extract the information about the food and 
calculate the total calories. 
First, you must extract the food info. It will always follow the same pattern rules:
It will be surrounded by "|" or "#" (only one of the two) in the following pattern: 
#{item name}#{expiration date}#{calories}#   or 
|{item name}|{expiration date}|{calories}|
The item name will contain only lowercase and uppercase letters and whitespace.
The expiration date will always follow the pattern: "{day}/{month}/{year}", where the day, month, and year will be 
exactly two digits long.
The calories will be an integer between 0-10000.
Calculate the total calories of all food items and then determine how many days you can last with the food you have. 
Keep in mind that you need 2000kcal a day.
Input / Constraints
You will receive a single string.
Output
First, print the number of days you will be able to last with the food you have:
"You have food to last you for: {days} days!"
The output for each food item should look like this:
"Item: {item name}, Best before: {expiration date}, Nutrition: {calories}"
'''

File: JavaScript projects/JavaScript Fundamentals/Final Exam/Exam_01/02_ad_astra.js
function solve(input){
    let regExp = /(#|\|)(?<food>[A-Za-z\s]+)\1(?<date>\d{2}\/\d{2}\/\d{2})\1(?<calories>(?:0|[1-9]\d{0,3}|10000))\1/g;
    let matches = input[0].matchAll(regExp);

    let food = [];
    let calories = 0;

    for (let match of matches){
        food.push([match.groups.food, match.groups.date, match.groups.calories]);
        calories += Number(match.groups.calories);
    }
    console.log(`You have food to last you for: ${Math.floor(calories / 2000)} days!`);
    for (let el of food){
        console.log(`Item: ${el[0]}, Best before: ${el[1]}, Nutrition: ${el[2]}`);
    }
}
solve([
'#Bread#19/03/21#4000#|Invalid|03/03.20||Apples|08/10/20|200||Carrots|06/08/20|500||Not right|6.8.20|5|'
]);
solve([ '$$#@@%^&#Fish#24/12/20#8500#|#Incorrect#19.03.20#450|$5*(@!#Ice Cream#03/10/21#9000#^#@aswe|Milk|05/09/20|2000|' ]);
solve(['Hello|#Invalid food#19/03/20#450|$5*(@']);